// @forestadmin/datasource-sequelize/dist/utils/model-to-collection-schema-converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sequelize_1 = require("sequelize");
const type_converter_1 = __importDefault(require("./type-converter"));
class ModelToCollectionSchemaConverter {
    static convertAssociation(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    association) {
        switch (association.associationType) {
            case sequelize_1.BelongsTo.name:
                return {
                    foreignCollection: association.target.name,
                    foreignKey: association.foreignKey,
                    foreignKeyTarget: association.targetKey,
                    type: 'ManyToOne',
                };
            case sequelize_1.BelongsToMany.name:
                return {
                    foreignCollection: association.target.name,
                    throughCollection: association.through.model.name,
                    originKey: association.foreignKey,
                    originKeyTarget: association.targetKey,
                    foreignKey: association.otherKey,
                    foreignKeyTarget: association.sourceKey,
                    type: 'ManyToMany',
                };
            case sequelize_1.HasMany.name:
                return {
                    foreignCollection: association.target.name,
                    originKey: association.foreignKey,
                    originKeyTarget: association.sourceKey,
                    type: 'OneToMany',
                };
            case sequelize_1.HasOne.name:
                return {
                    foreignCollection: association.target.name,
                    originKey: association.foreignKey,
                    originKeyTarget: association.sourceKey,
                    type: 'OneToOne',
                };
            default:
                throw new Error(`Unsupported association: "${association.associationType}".`);
        }
    }
    static convertAssociations(modelName, associations, logger) {
        const schemaAssociations = {};
        if (associations) {
            Object.entries(associations).forEach(([name, association]) => {
                try {
                    schemaAssociations[name] = this.convertAssociation(association);
                }
                catch (error) {
                    logger?.('Warn', `Skipping association '${modelName}.${name}' (${error.message})`);
                }
            });
        }
        return schemaAssociations;
    }
    static convertAttribute(attribute) {
        const sequelizeColumnType = attribute.type;
        const columnType = type_converter_1.default.fromDataType(sequelizeColumnType);
        const filterOperators = type_converter_1.default.operatorsForColumnType(columnType);
        const column = {
            columnType,
            filterOperators,
            type: 'Column',
            validation: [],
            isReadOnly: attribute.autoIncrement,
            isSortable: true,
        };
        if (attribute.allowNull === false &&
            !column.isReadOnly &&
            !attribute.defaultValue &&
            // eslint-disable-next-line no-underscore-dangle
            !attribute._autoGenerated) {
            column.validation.push({ operator: 'Present' });
        }
        if (attribute.primaryKey)
            column.isPrimaryKey = true;
        if (attribute.defaultValue !== null &&
            attribute.defaultValue !== undefined &&
            (columnType === 'Json' || typeof attribute.defaultValue !== 'object')) {
            column.defaultValue = attribute.defaultValue;
        }
        if (attribute.values) {
            column.enumValues = [...attribute.values];
        }
        return column;
    }
    static convertAttributes(modelName, attributes, logger) {
        const fields = {};
        Object.entries(attributes).forEach(([name, attribute]) => {
            try {
                fields[name] = this.convertAttribute(attribute);
            }
            catch (error) {
                logger?.('Warn', `Skipping column '${modelName}.${name}' (${error.message})`);
            }
        });
        return fields;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static convert(model, logger) {
        if (!model)
            throw new Error('Invalid (null) model.');
        return {
            actions: {},
            countable: true,
            fields: {
                ...this.convertAttributes(model.name, model.getAttributes(), logger),
                ...this.convertAssociations(model.name, model.associations, logger),
            },
            searchable: false,
            segments: [],
        };
    }
}
exports.default = ModelToCollectionSchemaConverter;
