// @forestadmin/datasource-sequelize/dist/utils/query-converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const sequelize_1 = require("sequelize");
const un_ambigous_1 = __importDefault(require("./un-ambigous"));
class QueryConverter {
    constructor(model) {
        this.model = model;
        this.dialect = this.model.sequelize.getDialect();
        this.col = this.model.sequelize.col;
        this.fn = this.model.sequelize.fn;
        this.where = this.model.sequelize.where;
    }
    asArray(value) {
        if (!Array.isArray(value))
            return [value];
        return value;
    }
    makeWhereClause(field, operator, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    value) {
        if (operator === null)
            throw new Error('Invalid (null) operator.');
        switch (operator) {
            case 'Blank':
                return {
                    [sequelize_1.Op.or]: [this.makeWhereClause(field, 'Missing'), { [sequelize_1.Op.eq]: '' }],
                };
            case 'Like':
                if (this.dialect === 'sqlite')
                    return this.where(this.col(field), 'GLOB', value.replace(/%/g, '*').replace(/_/g, '?'));
                if (this.dialect === 'mysql' || this.dialect === 'mariadb')
                    return this.where(this.fn('BINARY', this.col(field)), 'LIKE', value);
                return { [sequelize_1.Op.like]: value };
            case 'ILike':
                if (this.dialect === 'postgres')
                    return { [sequelize_1.Op.iLike]: value };
                if (this.dialect === 'mysql' || this.dialect === 'mariadb' || this.dialect === 'sqlite')
                    return { [sequelize_1.Op.like]: value };
                return this.where(this.fn('LOWER', this.col(field)), 'LIKE', value.toLocaleLowerCase());
            case 'NotContains':
                return {
                    [sequelize_1.Op.not]: this.makeWhereClause(field, 'Like', `%${value}%`),
                };
            case 'Equal':
                return { [sequelize_1.Op.eq]: value };
            case 'GreaterThan':
                return { [sequelize_1.Op.gt]: value };
            case 'In':
                return { [sequelize_1.Op.in]: this.asArray(value) };
            case 'IncludesAll':
                return { [sequelize_1.Op.contains]: this.asArray(value) };
            case 'LessThan':
                return { [sequelize_1.Op.lt]: value };
            case 'Missing':
                return { [sequelize_1.Op.is]: null };
            case 'NotEqual':
                return { [sequelize_1.Op.ne]: value };
            case 'NotIn':
                return { [sequelize_1.Op.notIn]: this.asArray(value) };
            case 'Present':
                return { [sequelize_1.Op.ne]: null };
            default:
                throw new Error(`Unsupported operator: "${operator}".`);
        }
    }
    /*
     * Delete and update sequelize methods does not provide the include options.
     * This method is developed to by pass this problem.
     */
    async getWhereFromConditionTreeToByPassInclude(conditionTree) {
        const include = conditionTree ? this.getIncludeFromProjection(conditionTree.projection) : [];
        const whereOptions = this.getWhereFromConditionTree(conditionTree);
        if (include.length === 0) {
            return whereOptions;
        }
        const keys = [...this.model.primaryKeyAttributes];
        const records = await this.model.findAll({ attributes: keys, where: whereOptions, include });
        const conditions = records.map(record => {
            const equals = keys.map(pk => new datasource_toolkit_1.ConditionTreeLeaf(pk, 'Equal', record.get(pk)));
            return datasource_toolkit_1.ConditionTreeFactory.intersect(...equals);
        });
        const union = datasource_toolkit_1.ConditionTreeFactory.union(...conditions);
        return this.getWhereFromConditionTree(union);
    }
    getWhereFromConditionTree(conditionTree) {
        if (!conditionTree)
            return {};
        const sequelizeWhereClause = {};
        if (conditionTree.aggregator !== undefined) {
            const { aggregator, conditions } = conditionTree;
            if (aggregator === null) {
                throw new Error('Invalid (null) aggregator.');
            }
            const sequelizeOperator = aggregator === 'And' ? sequelize_1.Op.and : sequelize_1.Op.or;
            if (!Array.isArray(conditions)) {
                throw new Error('Conditions must be an array.');
            }
            sequelizeWhereClause[sequelizeOperator] = conditions.map(condition => this.getWhereFromConditionTree(condition));
        }
        else if (conditionTree.operator !== undefined) {
            const { field, operator, value } = conditionTree;
            const isRelation = field.includes(':');
            const safeField = (0, un_ambigous_1.default)(this.model, field);
            sequelizeWhereClause[isRelation ? `$${safeField}$` : safeField] = this.makeWhereClause(safeField, operator, value);
        }
        else {
            throw new Error('Invalid ConditionTree.');
        }
        return sequelizeWhereClause;
    }
    computeIncludeFromProjection(projection, withAttributes = true) {
        return Object.entries(projection.relations).map(([relationName, relationProjection]) => {
            return {
                association: relationName,
                attributes: withAttributes ? relationProjection.columns : [],
                include: this.computeIncludeFromProjection(relationProjection, withAttributes),
            };
        });
    }
    getIncludeFromProjection(projection) {
        return this.computeIncludeFromProjection(projection, false);
    }
    getIncludeWithAttributesFromProjection(projection) {
        return this.computeIncludeFromProjection(projection);
    }
    getOrderFromSort(sort) {
        return (sort ?? []).map(({ field, ascending }) => {
            const path = field.split(':');
            return [...path, ascending ? 'ASC' : 'DESC'];
        });
    }
}
exports.default = QueryConverter;
