// @forestadmin/datasource-sequelize/dist/utils/type-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sequelize_1 = require("sequelize");
class TypeConverter {
    // TODO: Handle all ColumnTypes, not only PrimitiveTypes?
    static fromColumnType(columnType) {
        const dataType = TypeConverter.columnTypeToDataType[columnType];
        if (!dataType)
            throw new Error(`Unsupported column type: "${columnType}".`);
        return dataType;
    }
    static getColumnTypeFromDataType(dataType) {
        switch (dataType.key) {
            case sequelize_1.DataTypes.BOOLEAN.key:
                return 'Boolean';
            case sequelize_1.DataTypes.DATE.key:
            case sequelize_1.DataTypes.NOW.key:
                return 'Date';
            case sequelize_1.DataTypes.DATEONLY.key:
                return 'Dateonly';
            case sequelize_1.DataTypes.ENUM.key:
                return 'Enum';
            case sequelize_1.DataTypes.JSON.key:
            case sequelize_1.DataTypes.JSONB.key:
                return 'Json';
            case sequelize_1.DataTypes.BIGINT.key:
            case sequelize_1.DataTypes.DECIMAL.key:
            case sequelize_1.DataTypes.DOUBLE.key:
            case sequelize_1.DataTypes.FLOAT.key:
            case sequelize_1.DataTypes.INTEGER.key:
            case sequelize_1.DataTypes.MEDIUMINT.key:
            case sequelize_1.DataTypes.NUMBER.key:
            case sequelize_1.DataTypes.REAL.key:
            case sequelize_1.DataTypes.SMALLINT.key:
            case sequelize_1.DataTypes.TINYINT.key:
                return 'Number';
            case sequelize_1.DataTypes.CHAR.key:
            case sequelize_1.DataTypes.CITEXT.key:
            case sequelize_1.DataTypes.STRING.key:
            case sequelize_1.DataTypes.TEXT.key:
                return 'String';
            case sequelize_1.DataTypes.TIME.key:
                return 'Timeonly';
            case sequelize_1.DataTypes.UUID.key:
            case sequelize_1.DataTypes.UUIDV1.key:
            case sequelize_1.DataTypes.UUIDV4.key:
                return 'Uuid';
            default:
                throw new Error(`Unsupported data type: "${dataType}"`);
        }
    }
    static fromDataType(dataType) {
        if (dataType.key === sequelize_1.DataTypes.ARRAY.key) {
            const arrayDataType = dataType;
            return [TypeConverter.fromDataType(arrayDataType.type)];
        }
        return TypeConverter.getColumnTypeFromDataType(dataType);
    }
    static operatorsForColumnType(columnType) {
        if (Array.isArray(columnType)) {
            return new Set([...TypeConverter.baseOperators, 'In', 'IncludesAll', 'NotIn']);
        }
        switch (columnType) {
            case 'Boolean':
                return new Set([...TypeConverter.baseOperators]);
            case 'Uuid':
                return new Set([
                    ...TypeConverter.baseOperators,
                    'Contains',
                    'EndsWith',
                    'Like',
                    'StartsWith',
                ]);
            case 'Number':
                return new Set([
                    ...TypeConverter.baseOperators,
                    'GreaterThan',
                    'In',
                    'LessThan',
                    'NotIn',
                ]);
            case 'String':
                return new Set([
                    ...TypeConverter.baseOperators,
                    'In',
                    'Like',
                    'ILike',
                    'LongerThan',
                    'NotContains',
                    'NotIn',
                    'ShorterThan',
                ]);
            case 'Date':
            case 'Dateonly':
                return new Set([...TypeConverter.baseOperators, 'GreaterThan', 'LessThan']);
            case 'Enum':
                return new Set([...TypeConverter.baseOperators, 'In', 'NotIn']);
            case 'Json':
                return new Set([...TypeConverter.baseOperators]);
            default:
                return new Set();
        }
    }
}
exports.default = TypeConverter;
// TODO: Allow to differentiate NUMBER and INTEGER.
TypeConverter.columnTypeToDataType = {
    Boolean: sequelize_1.DataTypes.BOOLEAN,
    Date: sequelize_1.DataTypes.DATE,
    Dateonly: sequelize_1.DataTypes.DATEONLY,
    Enum: sequelize_1.DataTypes.ENUM,
    Json: sequelize_1.DataTypes.JSON,
    Number: sequelize_1.DataTypes.NUMBER,
    Point: null,
    String: sequelize_1.DataTypes.STRING,
    Timeonly: sequelize_1.DataTypes.TIME,
    Uuid: sequelize_1.DataTypes.UUID,
};
TypeConverter.baseOperators = [
    'Blank',
    'Equal',
    'Missing',
    'NotEqual',
    'Present',
];
