// @forestadmin/datasource-sequelize/dist/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const aggregation_1 = __importDefault(require("./utils/aggregation"));
const model_to_collection_schema_converter_1 = __importDefault(require("./utils/model-to-collection-schema-converter"));
const query_converter_1 = __importDefault(require("./utils/query-converter"));
const serializer_1 = __importDefault(require("./utils/serializer"));
class SequelizeCollection extends datasource_toolkit_1.BaseCollection {
    constructor(name, datasource, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    model, logger) {
        super(name, datasource);
        if (!model)
            throw new Error('Invalid (null) model instance.');
        this.model = model;
        this.col = this.model.sequelize.col;
        this.fn = this.model.sequelize.fn;
        this.aggregationUtils = new aggregation_1.default(this.model);
        this.queryConverter = new query_converter_1.default(this.model);
        const modelSchema = model_to_collection_schema_converter_1.default.convert(this.model, logger);
        this.enableCount();
        this.addFields(modelSchema.fields);
        this.addSegments(modelSchema.segments);
    }
    async create(caller, data) {
        const records = await this.model.bulkCreate(data);
        return records.map(record => serializer_1.default.serialize(record.get({ plain: true })));
    }
    async list(caller, filter, projection) {
        let include = this.queryConverter.getIncludeWithAttributesFromProjection(projection);
        if (filter.conditionTree) {
            include = include.concat(this.queryConverter.getIncludeFromProjection(filter.conditionTree.projection));
        }
        if (filter.sort) {
            include = include.concat(this.queryConverter.getIncludeFromProjection(filter.sort.projection));
        }
        const query = {
            attributes: projection.columns,
            where: this.queryConverter.getWhereFromConditionTree(filter.conditionTree),
            include,
            limit: filter.page?.limit,
            offset: filter.page?.skip,
            order: this.queryConverter.getOrderFromSort(filter.sort),
            subQuery: false,
        };
        const records = await this.model.findAll(query);
        return records.map(record => serializer_1.default.serialize(record.get({ plain: true })));
    }
    async update(caller, filter, patch) {
        await this.model.update(patch, {
            where: await this.queryConverter.getWhereFromConditionTreeToByPassInclude(filter.conditionTree),
            fields: Object.keys(patch),
        });
    }
    async delete(caller, filter) {
        await this.model.destroy({
            where: await this.queryConverter.getWhereFromConditionTreeToByPassInclude(filter.conditionTree),
        });
    }
    async aggregate(caller, filter, aggregation, limit) {
        let aggregationField = aggregation.field;
        if (aggregation.operation === 'Count' || !aggregationField) {
            aggregationField = '*';
        }
        else {
            aggregationField = this.aggregationUtils.quoteField(aggregationField);
        }
        const aggregationFunction = this.fn(aggregation.operation.toUpperCase(), this.col(aggregationField));
        const aggregationAttribute = [
            aggregationFunction,
            this.aggregationUtils.aggregateFieldName,
        ];
        const { groups, attributes: groupAttributes } = this.aggregationUtils.getGroupAndAttributesFromAggregation(aggregation.groups);
        let include = this.queryConverter.getIncludeFromProjection(aggregation.projection);
        if (filter.conditionTree) {
            include = include.concat(this.queryConverter.getIncludeFromProjection(filter.conditionTree.projection));
        }
        const order = this.aggregationUtils.getOrder(aggregationFunction);
        const query = {
            attributes: [...groupAttributes, aggregationAttribute],
            group: groups,
            where: this.queryConverter.getWhereFromConditionTree(filter.conditionTree),
            include,
            limit,
            order: [order],
            subQuery: false,
            raw: true,
        };
        const rows = await this.model.findAll(query);
        return this.aggregationUtils.computeResult(rows, aggregation.groups);
    }
}
exports.default = SequelizeCollection;
